modules/infra/datapath/control_output.c

```diff
@@ -0,0 +1,82 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2024 Christophe Fontaine
+
+#include "gr_control_output.h"
+
+#include <gr_control.h>
+#include <gr_graph.h>
+#include <gr_log.h>
+#include <gr_mbuf.h>
+
+#include <rte_ether.h>
+#include <rte_graph_worker.h>
+
+#include <stdlib.h>
+
+enum {
+	CONTROL_OUTPUT_ERROR,
+	EDGE_COUNT,
+};
+
+struct rte_ring *control_output_ring;
+
+static uint16_t control_output_process(
+	struct rte_graph *graph,
+	struct rte_node *node,
+	void **objs,
+	uint16_t n_objs
+) {
+	struct gr_control_output_msg msg;
+	struct rte_mbuf *mbuf;
+
+	for (unsigned i = 0; i < n_objs; i++) {
+		mbuf = objs[i];
+		msg.callback = control_output_mbuf_data(mbuf)->callback;
+		msg.mbuf = mbuf;
+
+		if (msg.callback == NULL
+		    || rte_ring_enqueue_elem(
+			       control_output_ring, &msg, sizeof(struct gr_control_output_msg)
+		       ) != 0) {
```

Comment:
hopefully, this will be less ugly with the public function exposed by the control plane.